riscv/src/bootloader.rs

```diff
@@ -9,6 +9,45 @@ pub const PAGE_SIZE_BYTES_LOG: usize = 10;
 // 32-Bit architecture -> 4 bytes per word
 pub const BYTES_PER_WORD: usize = 4;
 
+pub fn bootloader_preamble() -> String {
+    let mut preamble = r#"
+    // ============== bootloader-specific instructions =======================
+    // Write-once memory
+    let BOOTLOADER_INPUT_ADDRESS = |i| i;
+    let bootloader_input_value;
+    // Loads a value. If the cell is empty, the prover can choose a value.
+    instr load_bootloader_input X -> Y { {X, Y} in {BOOTLOADER_INPUT_ADDRESS, bootloader_input_value} }
+
+    // HACK: We should not use these columns & constraints and instead be able to use X.
+    // However, this seems to currently violate some constraints, because X is neither
+    // an input nor an output of the jump_to_bootloader_input_if_nonzero instruction.
+    col witness tmp_addr;
+    col witness tmp_addrInv;
+    col witness tmp_addrIsZero;
+    tmp_addrIsZero = 1 - tmp_addr * tmp_addrInv;
+    tmp_addrIsZero * tmp_addr = 0;
+    tmp_addrIsZero * (1 - tmp_addrIsZero) = 0;
+    (1 - instr_jump_to_bootloader_input_if_nonzero) * tmp_addr = 0;
+
+    // Sets the PC to the bootloader input at the provided index if it is nonzero
+    instr jump_to_bootloader_input_if_nonzero Y {
+        {Y, tmp_addr} in {BOOTLOADER_INPUT_ADDRESS, bootloader_input_value},
+        pc' = (1 - tmp_addrIsZero) * tmp_addr + tmp_addrIsZero * (pc + 1)
+    }
```

Comment:
I don't really understand what this new block is doing. Can you clarify here?

Reply:
I added a comment, maybe that helps? It reads from the bootloader input memory and otherwise behaves like `jump_dyn_if_nonzero`. Both needs to happen in one time step, because we have no registers to write any intermediate results to.

I think this is basically how all instructions would work if we didn't have persistent registers, right?